### ISSUE TITLE
Out-of-tree builds on Windows sometimes fail with mklink

### ISSUE THREAD

**Issue by sbutcher-arm** [ARMmbed/mbedtls#1496]
### Description
- Type: Bug 
- Priority: Minor

---------------------------------------------------------------
## Bug

**OS**  
linux|windows|

When building Mbed TLS, out-of-tree builds with CMake create symbolic links to various directories, to be able to access test files and scripts with a relative path.

Unfortunately, creation of links on Windows using `mklink` (see [CMakeLists.txt](https://github.com/ARMmbed/mbedtls/blob/fc458d0b9bbfe5e0cd6335eb083b7c15958a2610/tests/CMakeLists.txt#L118)) doesn't always work on all filesystems, and can fail in test VMs. 

As a fix, I propose a optional parameter to `cmake` to allow a local copy of the files to be made instead of a link, as after all, these are just data files for use by application software or scripts.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-2196

**Comment by yachoor** [comment]
It doesn't work with CMake generators based on MSYS or MinGW Makefiles (and probably more) on windows - file(TO_NATIVE_PATH ...) creates paths with "/" resulting in invalid switch error:
```
CMake Error at build/clion-cmake-debug/_deps/mbedtls-src/CMakeLists.txt:78 (message):
  Could not create symbolic link for:
  C:/Users/janusz.chorko/projects/baseApp/build/clion-cmake-debug/_deps/mbedtls-src/include/mbedtls
  --> Invalid switch - "Users".
```

**Comment by SergeySeroshtan** [comment]
> It doesn't work with CMake generators based on MSYS or MinGW Makefiles (and probably more) on windows - file(TO_NATIVE_PATH ...) creates paths with "/" resulting in invalid switch error:
> 
> ```
> CMake Error at build/clion-cmake-debug/_deps/mbedtls-src/CMakeLists.txt:78 (message):
>   Could not create symbolic link for:
>   C:/Users/janusz.chorko/projects/baseApp/build/clion-cmake-debug/_deps/mbedtls-src/include/mbedtls
>   --> Invalid switch - "Users".
> ```

CMake Bug: https://gitlab.kitware.com/cmake/cmake/issues/5939

The proposed solution is found here: https://github.com/scantailor/scantailor/blob/master/cmake/ToNativePath.cmake

**Comment by SergeySeroshtan** [comment]
PR with a fix -  #2286

**Comment by gyuri-szing** [comment]
Based on my investigations the functions executed when using the file(TO_NATIVE_PATH ) command are the same as the ones used to construct command lines dumped to makefiles. In an environment like MinGW or Cygwin these shall use forward slashes ("/"). This is because make will use bash or similar unix like shell as it's shell i these environments. As mklink is an internal command of cmd.exe it can only work with backward slashes ("\").
Thus the proper fix to me seems to be to get rid of TO_NATIVE_PATH, and instead convert the forward slashes to backward slashes whenever you use call cmd.exe.
